File: src/DashSpec.Core/Parsing/BlockSyntax.cs
namespace DashSpec.Core.Parsing;

internal enum BlockCloseStyle
{
    Brace,
    EndKeyword,
}

/// <summary>Dual block syntax: legacy <c>{ … }</c> or <c>… end kind</c> (ADR-0036).</summary>
internal static class BlockSyntax
{
    public static void BeginBlock(TokenReader reader)
    {
        reader.SkipNewlines();
        if (reader.IsAt(TokenKind.LBrace))
        {
            reader.Advance();
            reader.PushBlockClose(BlockCloseStyle.Brace);
            return;
        }

        reader.PushBlockClose(BlockCloseStyle.EndKeyword);
    }

    public static bool IsBlockEnd(
        TokenReader reader,
        string endKind,
        string? endId = null)
    {
        return reader.PeekBlockClose() switch
        {
            BlockCloseStyle.Brace => reader.IsAt(TokenKind.RBrace),
            _ => TryMatchEnd(reader, endKind, endId, consume: false),
        };
    }

    public static void ExpectBlockEnd(
        TokenReader reader,
        string endKind,
        string? endId = null)
    {
        if (reader.PeekBlockClose() is BlockCloseStyle.Brace)
        {
            reader.Expect(TokenKind.RBrace);
            reader.PopBlockClose();
            return;
        }

        if (!TryMatchEnd(reader, endKind, endId, consume: true))
        {
            var label = endId is null ? endKind : $"{endKind} {endId}";
            throw new DashSpecParseException($"Expected end {label}.");
        }

        reader.PopBlockClose();
    }

    public static bool TryMatchEnd(
        TokenReader reader,
        string expectedKind,
        string? expectedId,
        bool consume)
    {
        reader.SkipNewlines();
        var index = reader.SavePosition();
        if (!reader.TryKeyword("end"))
        {
            return false;
        }

        var actualKind = ReadEndKind(reader);
        if (!string.Equals(actualKind, expectedKind, StringComparison.OrdinalIgnoreCase))
        {
            reader.RestorePosition(index);
            return false;
        }

        string? actualId = null;
        if (reader.RawKind is TokenKind.Ident && !reader.IsOnNewline())
        {
            actualId = reader.ReadIdentSameLine();
        }

        if (expectedId is not null &&
            actualId is not null &&
            !string.Equals(actualId, expectedId, StringComparison.OrdinalIgnoreCase))
        {
            reader.RestorePosition(index);
            return false;
        }

        if (!consume)
        {
            reader.RestorePosition(index);
        }

        return true;
    }

    private static string ReadEndKind(TokenReader reader)
    {
        if (reader.TryKeyword("on"))
        {
            reader.ExpectKeyword("click");
            return "click";
        }

        return reader.ReadIdent();
    }
}

File: src/DashSpec.Modeling.Parse/BlockSyntax.fs
namespace DashSpec.Modeling.Parse

open System
open DashSpec.Modeling.Parse.Lexing
open DashSpec.Modeling.Core

/// Dual block syntax: legacy { … } or … end kind (ADR-0036).
module BlockSyntax =

    let beginBlock (reader: TokenReader) =
        reader.SkipNewlines()
        if reader.IsAt TokenKind.LBrace then
            reader.Advance()
            reader.PushBlockClose BlockCloseStyle.Brace
        else
            reader.PushBlockClose BlockCloseStyle.EndKeyword

    let private readEndKind (reader: TokenReader) =
        if reader.TryKeyword "on" then
            reader.ExpectKeyword "click"
            "click"
        else
            reader.ReadIdent()

    let tryMatchEnd (reader: TokenReader) (expectedKind: string) (expectedId: string option) (consume: bool) =
        reader.SkipNewlines()
        let saved = reader.SavePosition()
        if not (reader.TryKeyword "end") then false
        else
            let actualKind = readEndKind reader
            if not (String.Equals(actualKind, expectedKind, StringComparison.OrdinalIgnoreCase)) then
                reader.RestorePosition saved
                false
            else
                let actualId =
                    if reader.RawKind = TokenKind.Ident && not (reader.IsOnNewline()) then
                        Some(reader.ReadIdentSameLine())
                    else None
                match expectedId, actualId with
                | Some eid, Some aid when not (String.Equals(aid, eid, StringComparison.OrdinalIgnoreCase)) ->
                    reader.RestorePosition saved
                    false
                | _ ->
                    if not consume then reader.RestorePosition saved
                    true

    let isBlockEnd (reader: TokenReader) (endKind: string) (endId: string option) =
        match reader.PeekBlockClose() with
        | BlockCloseStyle.Brace -> reader.IsAt TokenKind.RBrace
        | _ -> tryMatchEnd reader endKind endId false

    let expectBlockEnd (reader: TokenReader) (endKind: string) (endId: string option) =
        match reader.PeekBlockClose() with
        | BlockCloseStyle.Brace ->
            reader.Expect TokenKind.RBrace
            reader.PopBlockClose()
        | _ when tryMatchEnd reader endKind endId true -> reader.PopBlockClose()
        | _ ->
            let label = match endId with Some id -> $"{endKind} {id}" | None -> endKind
            raise (DashSpec.Modeling.Core.DashSpecParseException($"Expected end {label}."))